design/design-doc-external-secret-stores.md

```diff
@@ -0,0 +1,416 @@
+
+##### API
+
+This will be controlled via annotations on managed or composite resources and no
+API change would be required. To propagate a connection secret to a supported
+secret store, users would need to add the following annotation on the resource:
+
+```
+runtime.crossplane.io/propagate-connection-secret-to: <supported-secret-store>
+```
+
+For example:
+
+```
+runtime.crossplane.io/propagate-connection-secret-to: vault
+runtime.crossplane.io/propagate-connection-secret-path: /path/prefix
+```
+
+In Crossplane runtime we will implement the following interface per secret store
+we want to support:
+
+```
+type ConnectionPropagator interface {
+	PropagateConnection(mg resource.Managed) *unstructured.Unstructured
+}
+```
+
+A Vault implementation of `ConnectionPropagator` would return a corresponding
+`GenericSecret` custom resource for a given managed resource in the 
+`PropagateConnection` function. When the annotation is found for a supported
+secret store in a managed resource, a custom resource of the secret store would
+be created in addition to the Kubernetes secret.
+
+We might consider adding support for deleting kubernetes secret after
+propagation with another annotation which would mean we will only use Kubernetes
+secrets as a temporary medium to propagate secret to the actual store:
+
+```
+runtime.crossplane.io/delete-secret-after-propagate: true
+```
+
+##### Example Deployment for Using Vault as Secret Store
+
+1. User deploys crossplane and providers including _provider-vault_.
+2. User configures _provider-vault_ with a `ProviderConfig`.
+3. User annotates a managed resource to propagate its connection details to 
+Vault.
+4. Once the managed resource is ready, a Vault `GenericSecret` custom resource
+created in addition to the connection details Kubernetes secret.
+5. _provider-vault_ stores secret in Vault.
+
+### Conclusion
+
+The problem of publishing connection details to external secret stores is a
+similar problem that Crossplane already solves but at a different layer.
+**Option 3** leverages most of the existing mechanisms but requires secrets to
+be stored in Kubernetes, at least for some time with a possible
+_delete-after-propagation_ option.
+
+When we want to go with an out-of-tree plugin based solution, we cannot use
+Crossplane’s existing extensibility solution, which is providers, if 
+communication over k8s api (via secrets) is not acceptable. **Option 2** tries
+to find a solution by introducing a new plugin solution while also trying to
+keep things simple by not introducing an additional authentication and
+authorization problem by suggesting a communication over unix domain sockets.
+However, since all providers and Crossplane itself require this functionality,
+deploying plugins as sidecars introduces another complexity which could be
+solved with a sidecar injecting mutating webhook.
+
+**Option 1**, as the most straightforward solution, suffers from the need for
```

Comment:
Comment from @negz 

Is there a possibility to start with option 1, but keep the option to evolve toward option 2 without breaking API changes? If so, that would be my preference.

My thinking is that building the plugin approach upfront is a little heavyweight if we find that maybe Secrets, Vault, and one or two cloud KMS services are all we need to support at first. It also would be subject to a less 'strongly typed' API. We could wait until it was becoming painful to support many backends in runtime before we invested in the plugin model. Perhaps we would find that we actually never got to that point and doing it all in tree worked pretty well.

I could imagine having 'plugin' be a distinct type of secret store that we could add later - something like https://gist.github.com/negz/b018eb03a55be2697549327d8490786c

Reply:
I think that sounds reasonable. 
Once we gather some more feedback, I will consider moving **option 2** and **option 3** under the **alternatives considered** section and extend **option 1** with what you're suggesting.

Reply:
I also like the Option 1 approach! I didn't know the option for plugins in Crossplane, I was thinking more to implement a specific provider for Vault or other secret managers.